be/src/exec/pipeline/scan/morsel.h

```diff
@@ -223,19 +226,49 @@ class FixedMorselQueue final : public MorselQueue {
     std::atomic<size_t> _pop_index;
 };
 
-class PhysicalSplitMorselQueue final : public MorselQueue {
+class SplitMorselQueue : public MorselQueue {
 public:
-    explicit PhysicalSplitMorselQueue(Morsels&& morsels, int64_t degree_of_parallelism, int64_t splitted_scan_rows)
+    SplitMorselQueue(Morsels&& morsels, int64_t degree_of_parallelism, int64_t splitted_scan_rows)
             : _morsels(std::move(morsels)),
               _num_original_morsels(_morsels.size()),
               _degree_of_parallelism(degree_of_parallelism),
               _splitted_scan_rows(splitted_scan_rows) {}
+    void set_tablets(const std::vector<TabletSharedPtr>& tablets) override { _tablets = tablets; }
+
+    void set_ticket_checker(const query_cache::TicketCheckerPtr& ticket_checker) { _ticket_checker = ticket_checker; }
+
+protected:
+    void _inc_num_splits(bool is_last) {
+        if (_ticket_checker == nullptr) {
+            return;
+        }
+        DCHECK(0 <= _tablet_idx && _tablet_idx < _tablets.size());
+        auto tablet_id = _tablets[_tablet_idx]->tablet_id();
+        _ticket_checker->enter(tablet_id, is_last);
+    }
+
+    ScanMorsel* _cur_scan_morsel() { return down_cast<ScanMorsel*>(_morsels[_tablet_idx].get()); }
+
+    const Morsels _morsels;
+    // The number of the morsels before split them to pieces.
+    const size_t _num_original_morsels;
+    const int64_t _degree_of_parallelism;
+    // The minimum number of rows picked up from a segment at one time.
+    const int64_t _splitted_scan_rows;
+
+    std::vector<TabletSharedPtr> _tablets;
+    std::atomic<size_t> _tablet_idx = 0;
```

Comment:
why only _tablet_idx is std::atomic?

Reply:
it is a field moved from PhysicalSplitMorselQueue, it is contended by multiple threads.